### ISSUE TITLE
Issue running tutorials 1 - 4

### ISSUE THREAD

**Issue by MRFaria** [pybee/toga#315]
When I try running the tutorials from the master branch on Ubuntu I get the following error:
from toga.style.pack import *
ImportError: No module named 'toga.style'

**Comment by danyeaw** [comment]
Thanks for submitting this issue. I should be able to help you troubleshoot, since I'm also on Ubuntu. Which version of Ubuntu are you running?

How did you install toga? There was just a very large change made to Toga, so the tutorials aren't going to work unless you clone the sources and pip install those.

**Comment by MRFaria** [comment]
I am runnning ubuntu 17.04
I cloned the master branch and installed by running 'pip install -e .' in the top level
I then also tried the normal 'pip install toga' and I run into the same issue.

Thanks

**Comment by danyeaw** [comment]
Can you make sure that toga-core and toga-gtk are also being installed from source by in addition to the command that you did, reperform it with src/core and src/gtk?

**Comment by MRFaria** [comment]
Well, when I try running 'pip install toga' in the top level directory I get the following:

Processing ./toga
Collecting toga-gtk (from toga==0.3.0.dev1)
  Using cached toga_gtk-0.2.15-py3-none-any.whl
Collecting toga-core>=0.2.15 (from toga-gtk->toga==0.3.0.dev1)
  Using cached toga_core-0.2.15-py3-none-any.whl
Collecting colosseum>=0.1.6 (from toga-core>=0.2.15->toga-gtk->toga==0.3.0.dev1)
  Using cached colosseum-0.1.6-py3-none-any.whl
Installing collected packages: colosseum, toga-core, toga-gtk, toga
  Running setup.py install for toga ... done
Successfully installed colosseum-0.1.6 toga-0.3.0.dev1 toga-core-0.2.15 toga-gtk-0.2.15

So it looks like toga-core and toga-gtk are being installed successfully

**Comment by danyeaw** [comment]
This is what I get when I type the same thing:

    $ pip3 install toga
    Requirement already satisfied: toga in myhome/Projects/toga
    Requirement already satisfied: toga-gtk in ./src/gtk (from toga)
    Requirement already satisfied: toga-core>=0.3.0.dev1 in ./src/core (from toga-gtk->toga)
    Requirement already satisfied: gbulb>=0.5.3 in myhome/.virtualenvs/toga/lib/python3.6/site-packages (from toga-gtk->toga)
    Requirement already satisfied: colosseum>=0.1.6 in myhome/.virtualenvs/toga/lib/python3.6/site-packages (from toga-core>=0.3.0.dev1->toga-gtk->toga)

Notice that in yours that it is installing toga-core-0.2.15 and toga-gtk-0.2.15 instead of 0.3.0.dev1
Please try typing:
    pip3 install -e src/core src/gtk --upgrade

**Comment by MRFaria** [comment]
Hmmn, this is weird. It is not letting me install toga-gtk 0.3.0.dev1

I get this complaint:
$ pip3 install -e . --no-cache-dir --upgrade
`Collecting` toga-core>=0.3.0.dev1 (from toga-gtk==0.3.0.dev1)
`Could not find a version that satisfies the requirement toga-core>=0.3.0.dev1 (from toga-gtk==0.3.0.dev1) (from versions: 0.2.0, 0.2.1, 0.2.2, 0.2.3, 0.2.4, 0.2.5, 0.2.6, 0.2.7, 0.2.8, 0.2.9, 0.2.10, 0.2.11, 0.2.12, 0.2.13, 0.2.14, 0.2.15)
`No matching distribution found for toga-core>=0.3.0.dev1 (from toga-gtk==0.3.0.dev1)

If I check what I have installed:
`$ pip3 list | grep toga
`toga-core (0.3.0.dev1, /home/mofro/Documents/toga/src/core)

**Comment by MRFaria** [comment]
Ok, I managed to get around the pip problems by installing with easy_install3 instead.
It looks like tutorial2 is broken and 1,3 and 4 run (but sizing of the widgets are weird).

[https://imgur.com/vH1o5mU](url)

**Comment by danyeaw** [comment]
Can you please check if you are still having issues after #314 was fixed?

**Comment by freakboy3742** [comment]
@MRFaria The "Allocating size to GtkScrolledWindow" error should be resolved if you update to the most recent code in master. 

As for the web widget being oddly positioned - there are some known layout issues on GTK; but that's a manifestation I haven't seen before (and can't reproduce on Fedora 27). What version of Fedora are you running?

**Comment by MRFaria** [comment]
I'm on ubuntu 17.04 and the "Allocating size" error has indeed been resolved. The weird widget positions remain. Although like before, some of the widgets like the ones in the toga-demo seem to be correctly positioned.

**Comment by anandr165** [comment]
I can confirm issue on Windows as well.
Moreover, the issue persists even if I clone the sources. 

This is the toga install i have from source:

$pip install -e toga
Obtaining file:///C:/Users/anandb/Downloads/Beeware/Toga-Handson/togareboot/toga
Requirement already satisfied: toga-winforms in c:\users\anandb\appdata\local\programs\python\python36\lib\site-packages (from toga==0.3.0.dev1)
Requirement already satisfied: toga-core>=0.2.15 in c:\users\anandb\appdata\local\programs\python\python36\lib\site-packages (from toga-winforms->toga==0.3.0.dev1)
Requirement already satisfied: pythonnet in c:\users\anandb\appdata\local\programs\python\python36\lib\site-packages (from toga-winforms->toga==0.3.0.dev1)
Requirement already satisfied: colosseum>=0.1.6 in c:\users\anandb\appdata\local\programs\python\python36\lib\site-packages (from toga-core>=0.2.15->toga-winforms->toga==0.3.0.dev1)
Installing collected packages: toga
  Running setup.py develop for toga
Successfully installed toga

C:\Users\anandb\Downloads\Beeware\Toga-Handson>python -m TemperatureConv
Traceback (most recent call last):
  File "C:\Users\anandb\AppData\Local\Programs\Python\Python36\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Users\anandb\AppData\Local\Programs\Python\Python36\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\anandb\Downloads\Beeware\Toga-Handson\TemperatureConv.py", line 2, in <module>
    from toga.style.pack import *
ModuleNotFoundError: No module named 'toga.style'


Moreover, 
even toga-demo seems broken.

Could you tell me what's the last stable version before the big change that went in recently so I can try that out to eliminate environment issues?

**Comment by freakboy3742** [comment]
@anandr165 I'm not sure toga-demo has ever worked on Windows - IIRC there's a couple of widgets still missing from the winforms support library.

However, based on the install log you've shown there, you're running a hybrid of toga versions. We're currently in the middle of a big refactor, and because there are several packages required, it's not just a matter of running "pip install ." from the root of the code checkout. You need to install toga-core, then toga-winforms, and then toga-demo, in that order, to prevent older versions from being accidentally pulled in via dependencies. 

The refactor is getting close to completion, so hopefully we'll be able to push updated versions shortly.

**Comment by sproutecks2125** [comment]
@freakboy3742 and @danyeaw 

I am having the same issue as the original poster. From what I have read through this issue post and replies windows may not be completely working correctly with Toga. I am almost completely new to python, GitHub, etc. Pretty much everything. I am learning as I run into problems for things I want to do while learning python, JS, C#.

Here is what I am getting (may not be formatted correctly so please don't kill me):

ModuleNotFoundError: No module named 'toga.style'

Stats on what I am running this on:
Windows 10 home
Visual Studio 2017 community (with virtual environment set up) - installing through VS2017 community.

I know this is probably not the way most would suggest, but when trying to navigate command line or PowerShell I kept running into problems.

The first tutorial worked and ran, the only issue was it was the button took up the entire window (again probably because I am using windows instead of the others).

I came across Toga via @freakboy3742's PyCon 2016 speech, and was excited and impressed with the possibilities.

If I just need to wait until after the refactor then that's what I will do. Just trying to learn to take my python code into GUI versus "print(abc)" etc.

Thank you for any help/suggestions

**Comment by jedie** [comment]
I ran into the same problem on ubuntu:
```
$ python -m run_calculator 
Traceback (most recent call last):
  File "/usr/lib/python3.5/runpy.py", line 184, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/jens/repos/RunCalculator-Env/run_calculator/run_calculator/__main__.py", line 1, in <module>
    from run_calculator.app import main
  File "/home/jens/repos/RunCalculator-Env/run_calculator/run_calculator/app.py", line 3, in <module>
    from toga.style import Pack
ImportError: No module named 'toga.style'
```

```
$ pip freeze
...
toga==0.2.15
toga-core==0.2.15
toga-gtk==0.2.15
...
```

**Comment by freakboy3742** [comment]
@jedie The tutorial you're looking at is for 0.3.0; if you've got 0.2.15 installed, that explains the error you've found.

**Comment by jedie** [comment]
I currently try to install all packages as editable from github. Because the current PyPi release is unuseable, because of the `ImportError: No module named 'toga.style'`

**Comment by freakboy3742** [comment]
@jedie You've been opening a *lot* of tickets recently; Can I please ask you to slow down?

Almost all these problems have been caused because of one fundamental misunderstanding about how to get Toga installed in the first place, and opening dozens of tickets is just creating work for us.

There is absolutely an issue to be addressed here. However, the approach you're taking to telling us about this issue is just creating more work for us, which is distracting from any effort to actually *fix* the problem.

The current PyPI release is *completely* usable - *if* you use the tutorial *for that version*. *That* is the underlying problem - people are finding various tutorials for 0.3.0 branch code, but using installation processes that install 0.2.15.

If you're using the 0.3.0 branch, you can use `pip install --pre toga` - which is [what the 0.3.0 tutorial recommends](http://toga.readthedocs.io/en/latest/tutorial/tutorial-0.html) - but this step *isn't* followed by most people trialling Toga.

The problem is then compounded when they assume installing "toga" from source will work - but it wont, because the toga package itself does nothing. The process of installing from source [is also documented](http://toga.readthedocs.io/en/latest/how-to/contribute.html), but people aren't finding and/or following those instructions.

Toga works fine - *if* you install it right. The catch is that for some reason, people either aren't finding or are ignoring the instructions that will help them install correctly. I don't say this to *blame* anyone - we've clearly got a problem with our documentation. 

Until we formally release 0.3.0, it's a problem we're going to have to deal with. However, opening a ticket in every repository you encounter isn't helping us address the problem.

**Comment by jedie** [comment]
Sorry for lots of open issues... I just try to use "pybee" for: https://github.com/jedie/RunCalculator/ and I just run into several bugs. If i found a Bug, then i report it. Shouldn't I do that? Should i use one issues? Is there a better place for this?

My first personal destination is Android and a web app.

I can't use 0.2.15 or the "pre" package, until a new release published to PyPi that fixed #412 and #413 ;)
So i use the install method described in https://toga.readthedocs.io/en/latest/how-to/contribute.html

**Comment by freakboy3742** [comment]
I don't want to discourage you from reporting bugs. I definitely appreciate that you're taking the time to experiment with Toga, and that you want to report the problems you've found. 

Bugs report like the one you gave for pybee/briefcase#110 are *really* helpful. They're specific, actionable - and in that case, came with a patch as well :-) (to be clear - having a patch is just an awesome bonus, not a requirement for submitting a ticket)

However, as far as I can make out, you're running:

* a development version of Toga, 
* on a codebase other than a tutorial
* on the two *least* mature platforms Toga supports (Android and Web).

I can tell you that using a development version of Toga *alone* is going to cause headaches - simply because Briefcase doesn't currently do a great job supporting that configuration (which is something that should be probably be opened as a feature request ticket). 

The combination of all three? Yeah - you're going to hit problems. :-) And keep in mind the version number as well - v0.3. That's us telling you that there are going to be problems.

If you found a bug while running the tutorial? Absolutely report it. But if you're experimenting on your own, individually reporting every minor problem you encounter isn't the most effective way to help us, because it doesn't help us narrow down the underlying problems. 

In cases like that, it's a lot more helpful to start a dialog - either by opening a single ticket as a focal point, or by jumping in the [chat room](https://gitter.im/pybee/general) and seeking guidance on what is a known problem, what never worked in the first place, and what is a genuinely a new bug.

In your specific case here, the solution is usually to specify a particular version in your `app_requires` (e.g., `toga-django==0.3.0.dev7`). That obviously doesn't help if you need to use a development version; but as I've already said, that's a configuration that Briefcase doesn't support at present (but should). My usual home-brewed solution is to run briefcase, then delete the toga install that briefcase added to the app_packages directory, and symlink the toga source directories in their place (you could probably also use `pip install --target`)

Again, I want to thank you for taking the time to tinker with Toga - your efforts are definitely appreciated. I'm just trying to ensure that your efforts aren't wasted, and we can make good use of the experience you're having.

**Comment by jedie** [comment]
Thanks for response.

> I can tell you that using a development version of Toga alone is going to cause headaches - simply because Briefcase doesn't currently do a great job supporting that configuration (which is something that should be probably be opened as a feature request ticket).

That's seems the point ;) And now i found https://github.com/pybee/briefcase/issues/73 this seems the be the right place to discuss this ;)

**Comment by danyeaw** [comment]
It sounds like the original issue that this was reporting is resolved by using:

    pip install --pre toga

If there are other issues with widget sizing or other problems please open a new issue.
